feat: workspace filter query supported in backend

Emyrk opened the pull request:
Continue of https://github.com/coder/coder/pull/2134

# What this does

Adds workspace filter query support in backend. Adds template_name query param support as well.

# Extra

- Make 1 `GetTemplatesWithFilter` rather than multiple sql queries with different params
- Make a unit test to keep the fake database methods to the set required (delete excessive methods)
- Add a `QueryParamParser` to handle parsing query params better.

# Future Work

Make the FE pass the query string as is to the BE. @f0ssel ?

f0ssel committed:
feat: add support for template in workspace filter

f0ssel committed:
get working

f0ssel committed:
make gen

Emyrk committed:
feat: Implement workspace search filter to support names

Emyrk committed:
Use new query param parser for pagination fields

Emyrk committed:
Fix fake db implementation

Emyrk committed:
Add unit test for parsing query params

Emyrk commented on coderd/httpapi/search.go:
```diff
@@ -0,0 +1,73 @@
+
+// WorkspaceSearchQuery takes a query string and breaks it into it's queryparams
+// as a set of key=value.
+func WorkspaceSearchQuery(query string) (map[string]string, error) {
```
Maybe I should move this to `workspaces.go`? 

I kinda want to make this more generic for when we add a `UserSearchQuery`, but felt I'll make it generic when we get that far.

Emyrk replied:
Moved. Might move again if we make other queries for users/templates page.

Emyrk committed:
Fix linting

Emyrk committed:
Fix search

Emyrk committed:
Maintain old behavior

Emyrk committed:
Linting

Emyrk committed:
Merge remote-tracking branch 'origin/main' into stevenmasley/workspace-template-search

Emyrk committed:
Remove excessive calls, use filters on a single query

Emyrk committed:
Remove unused code

Emyrk committed:
Add unit test to keep fake db clean

Emyrk commented on coderd/database/databasefake/databasefake_test.go:
```diff
@@ -0,0 +1,59 @@
+// functions. The fake database is a manual implementation, so it is possible
+// we forget to delete functions that we remove. This unit test just ensures
+// we remove the extra methods.
+func TestExactMethods(t *testing.T) {
```
This helps find which methods are not required on the fake. Just good for keeping that struct clean.

Emyrk committed:
Fix typo

kylecarbs commented on codersdk/workspaces.go:
```diff
@@ -218,9 +218,14 @@ func (c *Client) PutExtendWorkspace(ctx context.Context, id uuid.UUID, req PutEx
+	Template string `json:"template,omitempty"`
+	// Name will return partial matches
+	Name string `json:"name,omitempty"`
+	// FilterQuery supports a raw filter query string
+	FilterQuery string `json:"q,omitempty"`
```
This seems pretty non-standard to support both. When would this be used?

f0ssel replied:
I was on the fence with this as well. Github supports both a full query and an array of tags to list issues on their api for example. I'm in favor of removing the query params and add them back later if we need to. @Emyrk thoughts?

Emyrk replied:
@kylecarbs from a curl + cli perspective I figured it's easier to use queryparams.

From the UI, it's easier to build a single string. :shrug:.

Emyrk replied:
@kylecarbs It is better to parse the query param on the backend imo, since the backend controls the actual behavior of the search query. Makes the FE job easier.

But from an api perspective, I really just figured query params are easier. We use query params in other places, so if we only support `q` and not the others, we have different behavior depending on the endpoint.

Emyrk replied:
I have to convert the query to params regardless. So it's not really any more code.

kylecarbs replied:
It seems non-standard to have both from my perspective. What do you think about disallowing other filters in favor of the queryable format?

f0ssel commented on coderd/database/queries/templates.sql:
```diff
@@ -8,13 +8,39 @@ WHERE
+			organization_id = @organization_id
+		ELSE true
+	END
+	-- Filter by name, matching on substring
+	AND CASE
+		WHEN @name :: text != '' THEN
+			LOWER(name) LIKE '%' || LOWER(@name) || '%'
+		ELSE true
+	END
```
Do we need this? We only do substring matching on name as a special case, we probably don't need it for other query fields thought.

Emyrk replied:
I can drop it for now. I will keep the other field as `exact_name` though

Emyrk replied:
When we add these search bars on other pages, we might want it back.

Emyrk committed:
Drop like name search on template

Emyrk committed:
Fix linting

Emyrk committed:
Move WorkspaceSearchQuery to workspaces.go

Emyrk committed:
Search all templates with name

Emyrk committed:
Add more complex filter unit test

f0ssel reviewed (approved):
I think it's fine to support both since we already have that support, or just the query since it's more minimalist. We can always change this later so I'm not going to let the decision block this work.

Emyrk committed:
Fix unit test to not violate pg constraint

f0ssel commented on coderd/workspaces.go:
```diff
@@ -103,38 +103,60 @@ func (api *API) workspace(rw http.ResponseWriter, r *http.Request) {
+	}
+
+	if filter.OwnerUsername == "me" {
+		if !(filter.OwnerID == uuid.Nil || filter.OwnerID == apiKey.UserID) {
+			httpapi.Write(rw, http.StatusBadRequest, httpapi.Response{
+				Message: "Cannot set both \"me\" in \"owner_name\" and use \"owner_id\".",
+			})
+			return
+		}
```
I'm kinda hesitant to return 4xx errors in this route since it's powered by a filter. We can add support for validation error in the query bar on the dashboard, but right now we the behavior is we just take the first. 

Could we just drop ownerID? I think we use owner name on the app everywhere anyways?

Emyrk replied:
I'm ok with dropping `owner_id` for now :+1:

Emyrk replied:
I'm keeping it in the database filter because apiKey has userID, not their name. But I'll drop it from the query params.

Emyrk committed:
Drop owner_id from query params

Emyrk committed:
Remove unused code/params

Emyrk committed:
Merge remote-tracking branch 'origin/main' into stevenmasley/workspace-template-search

Emyrk committed:
Remove field from ts

kylecarbs commented on coderd/httpapi/queryparams.go:
```diff
@@ -0,0 +1,109 @@
+	}
+}
+
+// ValidationErrors is the set of errors to return via the API. If the length
+// of this set is 0, there was no errors.
+func (p QueryParamParser) ValidationErrors() []Error {
+	return p.errors
+}
```
I think `Errors` could be exposed on this struct to remove this function.

kylecarbs commented on coderd/httpapi/queryparams.go:
```diff
@@ -0,0 +1,109 @@
+	return p.errors
+}
+
+func (p *QueryParamParser) ParseInteger(r *http.Request, def int, queryParam string) int {
+	v, err := parse(r, strconv.Atoi, def, queryParam)
+	if err != nil {
+		p.errors = append(p.errors, Error{
+			Field:  queryParam,
+			Detail: fmt.Sprintf("Query param %q must be a valid integer (%s)", queryParam, err.Error()),
+		})
+	}
+	return v
+}
```
Because the struct is named `QueryParamParser`, we can remove `Parse` from these prefixes IMO.

kylecarbs commented on coderd/httpapi/queryparams_test.go:
```diff
@@ -0,0 +1,204 @@
+		testQueryParams(t, expParams, parser, parser.ParseInteger)
+	})
+
+	t.Run("UUIDArray", func(t *testing.T) {
```
Good tests! 👏

Emyrk commented:
> It seems non-standard to have both from my perspective. What do you think about disallowing other filters in favor of the queryable format?

If we don't allow `?owner=XXX` and only allow `?q="..."`, I still think we should keep the SDK at
```golang
type WorkspaceFilter struct {
	// Owner can be "me" or a username
	Owner string `json:"owner,omitempty"`
	// Template is a template name
	Template string `json:"template,omitempty"`
	// Name will return partial matches
	Name string `json:"name,omitempty"`
}
```

I'll need to add a function to convert these fields into the `q` string... but then our autogenerated type will be missing the `q` field for the frontend...

Do you prefer to do:
```golang
client.Workspaces(ctx, codersdk.WorkspaceFilter{
  FilterQuery: fmt.Sprintf("owner=%s name=%s", codersdk.Me, workspaceName),
}

// OR
client.Workspaces(ctx, codersdk.WorkspaceFilter{
  Owner: codersdk.Me,
  Name: workspaceName,
}
```

---


To me, it feels awkward to use the `q` param from within code. It really only makes sense from the UI, as the query has to be decomposed into it's elements anyway.

Emyrk committed:
Address PR comments

Emyrk commented:
Also github issues does support `labels` as a query param alonside the `q` filter string. So they do a mix as `?q=label:backend` and `?label=backend` is both supported.

But it seems some filter strings are not their own query params.

Emyrk committed:
Fix js test

Emyrk committed:
Fix unit test

kylecarbs commented on coderd/httpapi/queryparams.go:
```diff
@@ -0,0 +1,105 @@
+	return v
+}
+
+func parse[T any](r *http.Request, parse func(v string) (T, error), def T, queryParam string) (T, error) {
```
This name is too generic for what it does. I think `parseQueryParams` would be helpful.`

Emyrk replied:
Fair :+1:.

kylecarbs commented on coderd/httpapi/queryparams.go:
```diff
@@ -0,0 +1,105 @@
+	}
+}
+
+func (p *QueryParamParser) Integer(r *http.Request, def int, queryParam string) int {
```
This should be `Int` to follow the Go type naming idiom that's been adopted.

kylecarbs commented on coderd/httpapi/queryparams.go:
```diff
@@ -0,0 +1,105 @@
+	return v
+}
+
+func (p *QueryParamParser) UUIDArray(r *http.Request, def []uuid.UUID, queryParam string) []uuid.UUID {
```
`UUIDs` or `UUIDSlice` would be more appropriate. I'm personally in favor of `UUIDs`.

kylecarbs commented on coderd/workspacesearch.go:
```diff
@@ -0,0 +1,80 @@
+
+// WorkspaceSearchQuery takes a query string and breaks it into its queryparams
+// as a set of key=value.
+func WorkspaceSearchQuery(query string) (map[string]string, error) {
```
It seems odd that this would be exported from the `coderd` package. I'd personally prefer this be internal and tested from the endpoint itself to prevent any implementation mismatches. I'm open to thoughts here!

kylecarbs replied:
nit: I noticed `AuthorizeFilter` is exposed too, and it'd be nice to make that internal as part of this PR.

kylecarbs replied:
Similar comment for the `Authorize` function. It's exposed on the primary `API` struct, but it doesn't need to be from what it seems.

Emyrk replied:
I can unexport it, but I really much rather just test this pure function. Having to interpret what the results of this is based on the api feedback of the endpoint sound challenging. Testing this function as is is trivial.

kylecarbs commented on coderd/workspacesearch.go:
```diff
@@ -0,0 +1,80 @@
+// will search `template:"my name:name"`, which produces an empty list instead of
+// an error.
+// nolint:revive
+func splitElements(query string, delimiter rune, maintainQuotes bool) []string {
```
This is a pretty generically named function for being in the `coderd` package. Something like `splitQueryParameterByDelimiter` would be better from my POV.

kylecarbs commented on coderd/workspacesearch.go:
```diff
@@ -0,0 +1,80 @@
+package coderd
```
This being in a separate file could lead Coder's (😉) to expect the workspace searching route exists in this file. It'd be my preference for this to go in `workspaces.go` instead.

kylecarbs commented on coderd/workspaces_test.go:
```diff
@@ -867,3 +1067,11 @@ func mustLocation(t *testing.T, location string) *time.Location {
 	return loc
 }
+
+func must[T any](s T, err error) T {
+	if err != nil {
+		panic(err)
+	}
+
+	return s
+}
```
Because this is only used once, I'd rather us `require.NoError` it for now. I'm not opposed entirely, but I'd rather introduce this in a way that's consistent with the rest of the codebase, not just used in one spot.

Emyrk replied:
The `cryptorand` package should really have `MustIntn` and other funcs. Without that, you cannot use `cryptorand` inside struct initialization, like `testCases`.

`require.NoError` does not help in this case. I'll just used a fixed index.

kylecarbs replied:
Cryptographic randomness isn't guaranteed on Mac and Windows, which is why they were removed. It's not safe to use in production, and could cause a panic.

Emyrk replied:
It's in a unit test. Not production

Emyrk replied:
Maybe a `cryptorandtest` pkg... unsure. Just hurts readability to have to instantiate vars outside the `testCases` structs.

kylecarbs commented on coderd/workspacesearch.go:
```diff
@@ -0,0 +1,80 @@
+	// Because we do this in 2 passes, we want to maintain quotes on the first
+	// pass.Further splitting occurs on the second pass and quotes will be
+	// dropped.
+	elements := splitElements(query, ' ', true)
```
I think this function can be replaced by something like this: https://go.dev/play/p/-z3GWlF-NDJ.

It appears it's duplication a lot of logic from `strings.Split`, which seems sketchy to me.

Emyrk replied:
It is going to share a lot of code with strings.Split, it is just catching quotes blocks to prevent the misunderstanding in the comment above it.

I can switch to `FieldsFunc`. The only downside to that is that it doesn't trim the quotes, but I can do that after.

kylecarbs commented on coderd/workspacesearch.go:
```diff
@@ -0,0 +1,80 @@
+// WorkspaceSearchQuery takes a query string and breaks it into its queryparams
+// as a set of key=value.
+func WorkspaceSearchQuery(query string) (map[string]string, error) {
+	searchParams := make(map[string]string)
```
Since this function is strongly typed, it seems strange to return with an arbitrarily indexed map. Could we return `codersdk.WorkspaceFilter` instead?

Emyrk replied:
If I do that, then I need to do the parsing in the function too. 

I know you dislike that you can use query params to assign the `key=value` and use `q=key:value`. In the case it's shared, the code is just easier to reuse if I map the `WorkspaceSearchQuery` to a map[string]string. I can change it to `url.Values`.

---

If we drop supporting query params `?owner=me` then I could return a concrete type more easily.

Emyrk replied:
I just dropped query params except `q`, so this is now strongly typed.

Emyrk committed:
PR feedback

- Unexport workspace search query function
- move code into workspaces.go
- use strings.SplitFields

Emyrk committed:
Rename vague function name

Emyrk commented:
@kylecarbs I'll just drop non `q` query params. Not worth holding up the PR over, I'll make the SDK convert the struct to the filter string.

Emyrk committed:
WorkspaceSearchQuery now returns db filter

Emyrk committed:
Correct unit test and typescript

Emyrk committed:
Merge remote-tracking branch 'origin/main' into stevenmasley/workspace-template-search

Emyrk committed:
fmt

Emyrk committed:
Use !== over !=

Emyrk committed:
Fix js unit test

Emyrk committed:
Js linting

kylecarbs commented on coderd/database/queries/templates.sql:
```diff
@@ -8,13 +8,33 @@ WHERE
 	1;
 
--- name: GetTemplatesByIDs :many
+-- name: GetTemplatesWithFilter :many
```
nit: This is really just GetTemplates (and that's how we use it from our API)
